src/main/java/com/boardgo/oauth2/service/CustomOAuth2UserService.java

```diff
@@ -0,0 +1,59 @@
+package com.boardgo.oauth2.service;
+
+import com.boardgo.domain.mapper.UserInfoMapper;
+import com.boardgo.domain.user.entity.ProviderType;
+import com.boardgo.domain.user.entity.UserInfoEntity;
+import com.boardgo.domain.user.repository.UserRepository;
+import com.boardgo.oauth2.dto.OAuth2CreateUserRequest;
+import com.boardgo.oauth2.dto.OAuth2Response;
+import com.boardgo.oauth2.dto.OAuth2UserResponseFactory;
+import com.boardgo.oauth2.entity.CustomOAuth2User;
+import java.util.Objects;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
+import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
+import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
+import org.springframework.security.oauth2.core.user.OAuth2User;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CustomOAuth2UserService extends DefaultOAuth2UserService {
+
+    private final UserRepository userRepository;
+    private final UserInfoMapper userInfoMapper;
+
+    @Override
+    public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
+        OAuth2User oAuth2User = super.loadUser(userRequest);
+        return this.process(userRequest, oAuth2User);
+    }
+
+    private OAuth2User process(OAuth2UserRequest userRequest, OAuth2User oAuth2User) {
+        ProviderType providerType =
+                ProviderType.valueOf(
+                        userRequest.getClientRegistration().getRegistrationId().toUpperCase());
+        OAuth2Response oAuth2Response =
+                OAuth2UserResponseFactory.getOAuth2Response(
+                        providerType, oAuth2User.getAttributes());
+
+        UserInfoEntity userInfoEntity =
+                userRepository
+                        .findByEmailAndProviderType(oAuth2Response.getEmail(), providerType)
+                        .orElse(null);
+        if (Objects.isNull(userInfoEntity)) {
+            userInfoEntity =
+                    createUser(
+                            new OAuth2CreateUserRequest(oAuth2Response.getEmail(), providerType));
```

Comment:
```suggestion
        UserInfoEntity userInfoEntity =
                userRepository
                        .findByEmailAndProviderType(oAuth2Response.getEmail(), providerType)
                        .orElse(createUser(
                            new OAuth2CreateUserRequest(oAuth2Response.getEmail(), providerType)));
```
혹시 이렇게는 안되나요? (가능한지 몰라서..)

Reply:
음.. 가능은 한데, 제안주신 코드로 수정하면
**`findByEmailAndProviderType()` 에서 기존에 회원이 있을 경우에 update 에 대한 분기처리가 가능한가요?**
```java
else {
            userInfoEntity.updateEmail(oAuth2Response.getEmail());
        }
```

그런데 생각해보니 (다른 이야기지만) update 코드를 수정해야겠어요
email 이 변경될 경우 회원의 이메일 정보를 갱신하는건데, 어차피 email 을 변경했다면 `findByEmailAndProviderType()` 에서 조회되지 않을테니!
그래서 **소셜로그인에서 제공해주는 회원의 고유ID 번호를 저장하는건 어떤가요?**
카카오에서는 회원이 email 을 변경할 수 있다고 kakao developers 문의에서 본적이 있어서 회원의 고유ID 를 저장하면 이메일이 변경되어도 고유ID 로 조회해서 갱신할 수 있으니까요